sdks/python/apache_beam/typehints/decorators.py

```diff
@@ -265,23 +368,34 @@ def _unpack_positional_arg_hints(arg, hint):
   return hint
 
 
-def getcallargs_forhints(func, *typeargs, **typekwargs):
-  """Like inspect.getcallargs, but understands that Tuple[] and an Any unpack.
+def getcallargs_forhints(using_var_hints, func, *typeargs, **typekwargs):
+  """Like inspect.getcallargs, with support for declaring default args as Any.
+
+  In Python 2, understands that Tuple[] and an Any unpack.
+
+  Args:
+    using_var_hints: For variable length arguments, whether to expect the bound
```

Comment:
I have to admit, I find this boolean setting quite hard to follow, even looking where it's used (once True and once False). What is the overall goal we're trying to achieve with this flexibility?

Reply:
getcallargs_forhints is being passed two variations of values for variadic arguments and it needs to be able to tell them apart.
After calling `signature.bind()`, getcallargs_forhints sees a tuple or a dict (for VAR_POSITIONAL or VAR_KEYWORD), which might be wrapping a well formed hint for that argument or one or more hints that need to be combined into a Tuple or Dict.

For example, it might see a `VAR_POSITIONAL` with a value of `tuple(int, str)` and need to convert that to `Tuple[Union[int, str], ...]`, or it might see a value of `tuple(Tuple[Any, ...])` and just need to unwrap it to `Tuple[Any, ...]`.

Reply:
I think we might be able to auto-detect the variation in use.

For VAR_POSITIONAL:
- True if we got tuple(Tuple[FOO, ...]), otherwise False

For VAR_KEYWORD:
- True if we got tuple(Dict[FOO, BAR]), otherwise False